base/random/RNGs.jl

```diff
@@ -0,0 +1,424 @@
+    MersenneTwister(seed, state, zeros(Float64, MTCacheLength), MTCacheLength)
+
+"""
+    MersenneTwister(seed)
+
+Create a `MersenneTwister` RNG object. Different RNG objects can have
+their own seeds, which may be useful for generating different streams
+of random numbers.
+
+# Examples
+```jldoctest
+julia> rng = MersenneTwister(1234);
+
+julia> x1 = rand(rng, 2)
+2-element Array{Float64,1}:
+ 0.590845
+ 0.766797
+
+julia> rng = MersenneTwister(1234);
+
+julia> x2 = rand(rng, 2)
+2-element Array{Float64,1}:
+ 0.590845
+ 0.766797
+
+julia> x1 == x2
+true
+```
+"""
+MersenneTwister(seed) = srand(MersenneTwister(Vector{UInt32}(), DSFMT_state()), seed)
+
+function copy!(dst::MersenneTwister, src::MersenneTwister)
+    copy!(resize!(dst.seed, length(src.seed)), src.seed)
+    copy!(dst.state, src.state)
+    copy!(dst.vals, src.vals)
+    dst.idx = src.idx
+    dst
+end
+
+copy(src::MersenneTwister) =
+    MersenneTwister(copy(src.seed), copy(src.state), copy(src.vals), src.idx)
+
+==(r1::MersenneTwister, r2::MersenneTwister) =
+    r1.seed == r2.seed && r1.state == r2.state && isequal(r1.vals, r2.vals) &&
+    r1.idx == r2.idx
+
+
+### Low level API
+
+@inline mt_avail(r::MersenneTwister) = MTCacheLength - r.idx
+@inline mt_empty(r::MersenneTwister) = r.idx == MTCacheLength
+@inline mt_setfull!(r::MersenneTwister) = r.idx = 0
+@inline mt_setempty!(r::MersenneTwister) = r.idx = MTCacheLength
+@inline mt_pop!(r::MersenneTwister) = @inbounds return r.vals[r.idx+=1]
+
+function gen_rand(r::MersenneTwister)
+    dsfmt_fill_array_close1_open2!(r.state, pointer(r.vals), length(r.vals))
+    mt_setfull!(r)
+end
+
+@inline reserve_1(r::MersenneTwister) = (mt_empty(r) && gen_rand(r); nothing)
+# `reserve` allows one to call `rand_inbounds` n times
+# precondition: n <= MTCacheLength
+@inline reserve(r::MersenneTwister, n::Int) = (mt_avail(r) < n && gen_rand(r); nothing)
+
+# precondition: !mt_empty(r)
+@inline rand_inbounds(r::MersenneTwister, ::Type{Close1Open2}) = mt_pop!(r)
+@inline rand_inbounds(r::MersenneTwister, ::Type{CloseOpen}) =
+    rand_inbounds(r, Close1Open2) - 1.0
+@inline rand_inbounds(r::MersenneTwister) = rand_inbounds(r, CloseOpen)
+
+@inline rand_ui52_raw_inbounds(r::MersenneTwister) =
+    reinterpret(UInt64, rand_inbounds(r, Close1Open2))
+@inline rand_ui52_raw(r::MersenneTwister) = (reserve_1(r); rand_ui52_raw_inbounds(r))
+@inline rand_ui2x52_raw(r::MersenneTwister) =
+    rand_ui52_raw(r) % UInt128 << 64 | rand_ui52_raw(r)
+
+rand_ui10_raw(r::MersenneTwister) = rand_ui52_raw(r)
+rand_ui23_raw(r::MersenneTwister) = rand_ui52_raw(r)
+
+### generation
+
+#### floats
+
+@inline rand(r::MersenneTwister, ::Type{I}) where {I<:FloatInterval} =
+    (reserve_1(r); rand_inbounds(r, I))
+
+@inline rand(r::MersenneTwister, T::BitFloatType) = rand_generic(r, T)
+
+#### random integers
+
+@inline rand(r::MersenneTwister,
+             ::Type{T}) where {T<:Union{Bool,Int8,UInt8,Int16,UInt16,Int32,UInt32}} =
```

Comment:
this formatting looks pretty strange - isn't there a shorthand name for this union somewhere?